쥬스메이커 [STEP3] 쥬봉이, Kiseok

carti1108 opened the pull request:
@Tediousday93 
안녕하세요! Rowan! step3 PR 보냅니다!
잘부탁드립니다!

### 코드를 작성할 때 고민되었던 점
- view state method 호출 부분 modal이라 viewWillAppear가 안나타나서 
-> 모달 스타일을 fullscreen으로 변경

- stepper로 재고 변경시 (-)가 안되거나 0이하로 내려가는 경우 발생
-> stepper별 초기값을 설정, 해당 부분을 label.text로 만들어줌

- fruit재고 변경하는 방법 : 프로퍼티 직접 변경 vs 메서드로 변경
-> view controller가 fruit재고 직접 변경하지 않도록 fruitStore에 재고 변경 메서드 추가

<br>

### 해결이 되지 않은 점 & 조언을 얻고 싶은 부분
- typealias를 사용하려 하는데 일단은 파일을 새로 추가하여 그 부분에서 선언을 해주고 사용했습니다.
이렇게 사용하는게 맞는지 궁금합니다. 
-> [Fruit: Int]를 FruitStock으로 typealias

jyubong committed:
feat: stepper 연결, stepper 메서드 구현, 데이터 전달 구현

stepper value 초기값 설정하는 메서드 추가
stepper 클릭시 label 값 변경
fruitStore 프로퍼티 값 변경 메서드 추가

carti1108 committed:
refactor: stockViewController autolayout 설정

jyubong committed:
refactor: 오토레이아웃 수정, fruits 재고 수정 메서드 로직 수정

jyubong committed:
refactor: [Fruit: Int] -> FruitStock type alias

carti1108 committed:
style: 파일 생성자 주석 변경

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -31,6 +31,12 @@ final class JuiceMakerViewController: UIViewController {
+        super.viewWillAppear(animated)
+        
+        setupUI()
+    }
```
재고 변경 화면에서 돌아왔을 때 label을 업데이트 하고 있군요 👍
하지만 가장 처음 앱을 실행했을 때 viewDidLoad와 viewWillAppear가 모두 호출되기 때문에 setupUI가 결국 2회 호출되게 됐네요 💭

jyubong replied:
setUpUI가 중복되어 viewWillAppear에서만 호출되도록 변경하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -110,6 +116,7 @@ final class JuiceMakerViewController: UIViewController {
             return
         }
         
+        stockNavigationController.modalPresentationStyle = .fullScreen
```
viewWillAppear가 호출될 수 있도록 하기 위해 modalPresentationStyle을 변경하셨군요!
fullScreen 모달을 사용하지 않더라도 label 업데이트 메서드를 호출할 수 있는 방법으로 `delegation 패턴`이 있습니다.
머지가 가능할 정도까지 수정이 된 이후에도 시간이 남는다면 delegation 패턴에 도전해봐도 좋을 것 같네요 😊

carti1108 replied:
수정사항 모두 변경 후 머지 직전에 한번 도전해보겠습니다!

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -15,23 +15,55 @@ final class StockChangeViewController: UIViewController {
+        pineappleStepper.value = Double(fruitStore.fruits[.pineapple] ?? .zero)
+        kiwiStepper.value = Double(fruitStore.fruits[.kiwi] ?? .zero)
+        mangoStepper.value = Double(fruitStore.fruits[.mango] ?? .zero)
+    }
```
처음 stepper의 value를 설정하고 있으므로 이름을 `setStepperValue`라고 하는건 어떨까요?
이렇게 같은 종류의 UI component를 다수 사용할 경우 Outlet Collection을 활용해볼 수도 있을 것 같아요!

jyubong replied:
함수 이름을 setStepperValue로 변경하였습니다.
`Outlet Collection`을 공부한 뒤 label과 stepper outlet을 collection으로 적용하였습니다.
collection으로 설정시 array로 만들어지는데, index의 경우 연결 순서로 정해지나 `순서 보장을 위해 tag를 설정하여 tag순으로 정렬되도록 didset을 구현`해보았습니다. (ex. stepperCollection.sort { $0.tag < $1.tag } )

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -15,23 +15,55 @@ final class StockChangeViewController: UIViewController {
+        bananaStockLabel.text = Int(bananaStepper.value).description
+        pineappleStockLabel.text = Int(pineappleStepper.value).description
+        kiwiStockLabel.text = Int(kiwiStepper.value).description
+        mangoStockLabel.text = Int(mangoStepper.value).description
```
description 프로퍼티를 활용하셨네요!
혹시 사용하기 전에 [CustomStringConvertible](https://developer.apple.com/documentation/swift/customstringconvertible)이라는 프로토콜에 대해 알아보셨나요?
공식문서를 참고해서 권장되는 활용방식이 있는지 찾아봅시다 💪

carti1108 replied:
Calling this property directly is discouraged.
공식문서에서 description을 직접 호출하는 것은 권장하지 않는다는 부분을 보고 StockChangeViewController 에서 사용되는 것들을 모두 변경해주었습니다!

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -15,23 +15,55 @@ final class StockChangeViewController: UIViewController {
+    
+    @IBAction func stepperTapped(_ sender: UIStepper) {
+        setupUI()
+    }
```
stepper 하나를 누르면 전체 레이블이 업데이트 될 것 같아요.
관련된 항목만 업데이트될 수 있도록 수정해볼까요?

jyubong replied:
`sender.tag를 활용`하여 관련 항목만 업데이트될 수 있도록 수정하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -15,23 +15,55 @@ final class StockChangeViewController: UIViewController {
+        fruitChange[.mango] = Int(mangoStepper.value)
+        
+        fruitStore.changeStock(by: fruitChange)
+    }
```
ViewController가 FruitStock(dictionary)에 대해 알 필요가 있을까요?? 🤔
dictionary를 관리하는 기능은 FruitStore에 있는 것이 좋다고 생각됩니다.

fruitStore에 추가한 재고 변경 메서드가 과일 하나의 재고를 변경하도록 만들면 좋을 것 같아요.
Fruits의 CaseIterable 프로토콜 채택으로 제공되는 allCases를 활용할 수 있게 될 것 같습니다.

carti1108 replied:
현재 메소드에서는 fruitStore의 changeStock메소드를 호출해서 Fruit의 allcases를 활용하여 fruit와 갯수만 전달인자로 보내도록 하였습니다!

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStock.swift:
```diff
@@ -0,0 +1,8 @@
+//  Created by 쥬봉이, Kiseok on 2023/09/21.
+//
+
+typealias FruitStock = [Fruit: Int]
```
typealias를 활용해보셨군요 👍

Dictionary보다는 tuple로 만들면 어떨까요?
새로 만든 메서드의 파라미터 이름이 fruitChange이기 때문에 이와 맞춰주면 좋겠네요!

현재 별칭을 ViewController에서 사용하게 되면 ViewController에 Dictionary 생성 및 element 추가 코드가 생기게 됩니다.
이런 기능은 우리가 만든 FruitStore가 하면 좋겠죠?!

jyubong replied:
[Fruit: Int]가 많이 사용되어 typealias로 사용해보았습니다.
사용해본 것에 의의를 두고, 주신 피드백에 더해 공부해서 다음에 적절한 곳에 활용해보겠습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,15 +8,15 @@
 class FruitStore {
     static let shared = FruitStore(initialStock: 10)
-    private(set) var fruits: [Fruit: Int] = [:]
+    private(set) var fruits: FruitStock = [:]
```
Dictionary같은 경우는 이미 축약형 타입을 사용하고 있기 때문에 따로 별칭을 지정하지 않아도 좋을 것 같아요 👀

jyubong replied:
typealias를 모두 삭제해주었습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -26,10 +26,16 @@ class FruitStore {
+        for (fruit, quantity) in fruitChange {
+            fruits.updateValue(quantity, forKey: fruit)
+        }
+    }
```
FruitStore 외부에서는 fruits dictionary를 관리하는 기능을 메서드로 호출할 수 있게끔 하기 위해
파라미터로 Dictionary를 직접 받기 보다는 Fruit 타입과 변경할 재고수량 Int타입을 받는 것이 적절하다고 생각됩니다.

2가지 개선 방법이 생각나네요. 💭

1. fruitChange를 tuple 타입으로 한다.
2. 변경할 과일(fruit)과 변경할 수량(count)을 따로 파라미터로 받는다.

만약 지금처럼 Dictionary를 전달받게 되면 이렇게 for문을 돌기보단 직접 fruits에 할당하는 것이 간단하겠죠?!

jyubong replied:
변경할 과일(fruit)과 변경할 수량(quantity)을 따로 파라미터로 받는 방식으로 변경하여 활용하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -15,23 +15,55 @@ final class StockChangeViewController: UIViewController {
+    }
+    
     @IBAction func doneButtonTapped(_ sender: Any) {
+        changeStock()
         self.presentingViewController?.dismiss(animated: true)
```
presentingViewController는 누구일까요? 💭

jyubong replied:
presentingViewController는 StockChangeViewController를 present한 `JuiceMakerViewController`를 나타냅니다.

Tediousday93 replied:
이전 화면에서 어떤 인스턴스를 present 했는지 확인해봅시다~!

jyubong replied:
UINavigationController를 present 하였습니다.

Tediousday93 reviewed (changes requested):
기석, 쥬봉이 안녕하세요 ☺️
step3 고생 많으셨습니다.

프로젝트 시간이 많이 남았기 때문에 이번 PR은 빠른 진행보다는
그 동안 챙기지 못했던 부분도 되돌아보는 시간을 가졌으면 좋겠습니다. 💪

디스코드에 모여서 같이 의견을 나눠봐도 좋을 것 같네요😄

</br>

### ✅ 고민했던 점
**view state method**
* modalPresentationStyle에 대해 찾아보셨겠네요 👍
* viewWillAppear가 호출되지 않아도 JuiceMakerView의 label들을 업데이트할 수 있는 방법이 있습니다. 이건 코멘트를 확인해주세요!

**stepper**
* 스토리보드에서 stepper.value의 min, max를 설정해줄 수 있죠!
* label.text와 stepper.value를 연결해주신 부분 👍👍👍

**과일 재고 변경 방법**
* ViewController가 재고를 직접 변경하지 않도록 하는 접근 방법은 정말 좋습니다 👍
* 하지만 현재 상태는 ViewController에서 fruits(Dictionary)를 직접 만들어서 전달해주기 때문에 FruitStore의 기능을 직접 수행하고 있다고 생각되네요.

</br>

### ✅ 해결이 되지 않은 점 / 조언을 얻고 싶은 부분
**typealias**
* 타입 별칭을 사용해보려는 시도 좋습니다. 말 그대로 별칭이기 때문에 모든 같은 타입을 별칭으로 바꾸지 않아도 됩니다.😊
* 이 별칭을 직접적으로 사용하는 타입 내부에 선언해주는게 좋습니다.💭 전역적인 별칭은 사용에 혼란을 야기할 수도 있을 것 같아요!
* 개인적으로 Dictionary의 경우 이미 축약형으로 타입을 표현하고 있기 때문에 따로 별칭을 만들어줄 필요는 없다고 생각합니다. 🤔

</br>

### 🔸 그 외 피드백
**AutoLayout**
* 재고 변경 화면의 오토레이아웃을 처음 잡아보실텐데 아주 잘 잡아주셨네요!
* StackView의 활용에서 조금은 아쉬운 부분이 있습니다. 야곰닷넷의 오토레이아웃 강의를 반복 학습하시는 것을 추천드립니다👍
* 이번에 구현하신 화면의 레이아웃은 잘 잡혀있기 때문에 오토레이아웃 수정보다는 코드 개선에 집중해봅시다! 👏

carti1108 committed:
refactor: stockViewController, FruitStore 메소드 수정

stockViewController의 iboutlet을 collection으로 변경, 메소드 수정
FruitStore의 changeStock메소드 수정

jyubong committed:
refactor: JuiceMakerViewController IBOutlet, setUpUI메서드 수정, stockViewcontroller stepper tapped 메서드 수정, typealias 삭제

jyubong committed:
docs: Update README.md

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -8,11 +8,11 @@ import UIKit
-    @IBOutlet weak var kiwiStockLabel: UILabel!
-    @IBOutlet weak var mangoStockLabel: UILabel!
+    @IBOutlet var labelCollection: [UILabel]! {
+        didSet {
+            labelCollection.sort { $0.tag < $1.tag }
+        }
```
didSet을 활용해서 tag에 맞게 정렬해주셨군요 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -27,16 +27,18 @@ final class JuiceMakerViewController: UIViewController {
-        pineappleStockLabel.text = String(fruitStore.fruits[.pineapple] ?? .zero)
-        kiwiStockLabel.text = String(fruitStore.fruits[.kiwi] ?? .zero)
-        mangoStockLabel.text = String(fruitStore.fruits[.mango] ?? .zero)
+        Fruit.allCases.enumerated().forEach { (index, fruit) in
+            labelCollection[index].text = String(fruitStore.fruits[fruit] ?? .zero)
+        }
```
`enumerated`와 `forEach` 를 잘 활용하셨군요 👍
훨씬 간단해졌네요!

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -9,29 +9,57 @@ import UIKit
         setupUI()
     }
     
+    private func setStepperValue() {
+        Fruit.allCases.enumerated().forEach { (index, fruit) in
+            stepperCollection[index].value = Double(fruitStore.fruits[fruit] ?? .zero)
+        }
+    }
```
👍👍👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -9,29 +9,57 @@ import UIKit
+    }
+    
     @IBAction func doneButtonTapped(_ sender: Any) {
+        changeStock()
```
stepper를 눌렀을 때 먼저 뷰에서 보여주고 화면이 dismiss될 때만 저장하도록 해주셨네요 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -9,29 +9,57 @@ import UIKit
+    
+    private func changeStepperValueToInt(at index: Int) -> Int {
+        Int(stepperCollection[index].value)
+    }
```
반복되는 코드를 메서드로 분리해주신 점 👍👍
extension을 통해 Double 타입에 `toInt`라는 메서드를 만들어도 괜찮을 것 같네요 😊
이미 잘 작성해주셨기 때문에 여기는 더 이상 수정하지 않으셔도 됩니다~

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -11,7 +11,7 @@ class FruitStore {
     
     private init(initialStock: Int) {
-        for fruit in Fruit.allCases {
+        Fruit.allCases.forEach { fruit in
```
후행클로저에 단축 인자이름을 활용해서 간단히 나타낼 수도 있습니다~!
```suggestion
        Fruit.allCases.forEach { self.fruits[$0] = initialStock }
```

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -27,9 +27,13 @@ class FruitStore {
     
     func decreaseStock(of recipe: [Fruit: Int]) {
-        for (fruit, quantity) in recipe {
+        recipe.forEach { (fruit, quantity) in
```
for문을 전부 고차함수로 바꿔주셨군요 👍

추가로 고려해볼 사항을 남겨드리겠습니다!
이전 스텝에서는 메서드 분리에 중점을 두기 위해서 넘어갔던 부분이에요.
`StockError`를 던져주는 주체는 누가 되는 것이 좋을까요? 💭

JuiceMaker는 사실 FruitStore에게 재고를 줄여달라고만 부탁하면 되고 재고 확인은 FruitStore가 하는 것이 좋다고 생각됩니다.
decreaseStock 내부에서 재고 확인을 하도록 하면 해결되겠죠?!
그럼 오류를 던져주는 객체도 적절하게 변경될 것 같네요.

또한, 아래 `changeStock(of:by:)`와 마찬가지로 하나의 과일에 해당하는 재고만 변경하도록 하여
호출부분에서 반복문을 활용하는 것이 좋아보입니다 🤔 (`isValidStock`도 마찬가지입니다~!)
FruitStore가 Juice의 recipe를 꼭 알아야 할 필요는 없으니 말이죠 😊

jyubong replied:
딸바주스, 망키주스와 같이 여러 과일을 사용할 경우 두 과일 모두 재고가 있는지 확인 후에 이상이 없을 경우 재고를 감소해야합니다.
JuiceMaker에서는 과일을 하나씩 전달하고 fruitStore에서 재고확인과 재고 감소를 할 경우 저희의 지식으로는 위의 로직을 구현하기 어렵습니다🥲
그래서 recipe을 주는 것보다 재고를 감소시킬 과일을 한번에 fruitStore로 전달한다는 느낌으로 현 코드는 유지하되 error 이름을 JuiceMakerError로 변경하여 error를 던지는 주체를 변경해주었습니다.🫶🏻 감사합니다.🫶🏻🫶🏻

Tediousday93 reviewed (changes requested):
전반적으로 정말 잘 수정해주셨어요 👏👏👏
피드백에 없던 추가 리팩토링까지 👍

아직 시간이 여유로우니 이번 피드백이 수정 완료되면 delegation 패턴까지 활용해봅시다 😊

jyubong committed:
refactor: fruitStore init 수정, juiceMaker와 fruitStore 메서드 로직 수정

carti1108 committed:
refactor:Error이름 변경, dismiss 수정

jyubong committed:
feat: Delegate protocol 추가

Tediousday93 commented:
고생 많으셨습니다! 👍
다음 프로젝트도 화이팅입니다 💪
